src/tools/perf/lib/ucp_tests.cc

```diff
@@ -725,6 +725,40 @@ class ucp_perf_test_runner {
         wait_send_window(m_max_outstanding);
     }
 
+    bool use_ack() const
+    {
+        /* TODO: daemon does not support bi-directional flow */
+        if (m_perf.params.ucp.is_daemon_mode) {
+            return false;
+        }
+
+        return (CMD == UCX_PERF_CMD_TAG) || (CMD == UCX_PERF_CMD_TAG_SYNC) ||
+               (CMD == UCX_PERF_CMD_AM);
+    }
+
+    void send_ack(ucp_ep_h ep, void *buffer, ucp_datatype_t datatype,
+                  uint64_t remote_addr, ucp_rkey_h rkey)
+    {
+        if (!use_ack()) {
+            return;
+        }
+
+        send(ep, buffer, 1, datatype, 0, remote_addr, rkey, false);
+        wait_send_window(m_max_outstanding);
+    }
+
+    void recv_ack(ucp_worker_h worker, ucp_ep_h ep, void *buffer,
+                  ucp_datatype_t datatype)
+    {
+        if (!use_ack()) {
+            return;
+        }
+
+        flush();
```

Comment:
why is flush needed?

Reply:
Right, normally we shouldn't need it, as we wait for receiver completion anyway, so removing it